include/api/CDataFrameAnalysisSpecification.h

```diff
@@ -15,14 +15,26 @@
 #include <rapidjson/fwd.h>
 
 #include <cstddef>
+#include <functional>
 #include <memory>
 #include <string>
 #include <thread>
 #include <vector>
 
+#ifdef HANDLE_FATAL_ERROR
+#undef HANDLE_FATAL_ERROR
+#endif
+#define HANDLE_FATAL_ERROR(handler, message)                                   \
```

Comment:
It may be better to merge this functionality into `LOG_AND_EXIT` in `CLogger.h` so that a customised exit call can be used _everywhere_ where either `LOG_AND_EXIT` or `HANDLE_FATAL_ERROR` is currently used.

I think it's a bit dodgy at the moment that there are some parts of the code where `LOG_AND_EXIT` is called in code that could be run by unit tests.  If one of these errors occurs in a unit test then we'll exit the test suite without writing the test results file and won't run any subsequent tests.

If `CLogger` had a method to register the default exit function then it would be possible to change the constructor of `ml::test::CTestRunner` so that it registered an exit function that did a `CPPUNIT_FAIL` instead of `std::exit`.  The default default exit function on construction of `CLogger` could still be `std::exit`.

We'd want the individual tests that override the fatal error handler to just collect the fatal messages, so maybe there is a need to keep `LOG_AND_EXIT` and `HANDLE_FATAL_ERROR` separate, but either way, I think a way should be found to make `LOG_AND_EXIT` call `CPPUNIT_FAIL` rather than `std::exit` if a fatal error occurs in a unit test.

Reply:
That sounds sensible. The remaining calls really shouldn't crop up in the unit tests, but your suggestion is neater and I like the idea of having a global override tied to the logger anyway.

Reply:
I've made [this](https://github.com/elastic/ml-cpp/pull/367/commits/52b23d77bcc92724b07e4c3c60c5068994c18e0a) change. This really opens up the follow on task to unify this with LOG_ABORT, but I don't want to make that change now. I'll raise an issue for that work.